### PR TITLE
Enforce the use of a single string formatting mechanism for translation source strings

### LINKED ISSUE #7111: Enforce the use of a single string formatting mechanism for translation source strings
### Is your proposal related to a problem?

Python has numerous ways of formatting strings which are all surfaced differently within Transifex.

Across the codebase, translation strings use different formatting: `{}`, `{0}`, `{named}`, `%s` and `%(named)s`. This can not only be confusing for new translation but all those tags aren’t treated equally in Transifex.

In addition to the different formatting styles, there is a case to be made for positional placeholders (i.e. {}, {0} and %s) which are problematic as it does not allow the target language to reorder them.

### Describe the solution you'd like

Enforce named and old style formatting:

```python
# In Python
_("Welcome to the %(site_name)s Wagtail CMS") % {"site_name": request.site.site_name}
```

```
<!-- In Templates -->
{% blocktrans %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}
```

### Additional context

Formatting Option | Screenshot | Notes
-- | -- | --
{}, {0} and {named} |Brackets_Positional_Empty  Brackets_Positional_Numbers Brackets_Named Brackets_Missing| There is no special treatment about them.When the name of a placeholder matches a glossary term, it displays an unrelevant warning.
%s and %(named)s |PercentS_Positional PercentS_Named PercentS_Question| The placeholders are clickable, it is clear they are not meant to be translated.Note that there were questions around the use of %s, but it was raised before placeholders were made clickable, so it probably isn’t an issue any more.
Missing {}, {0} or {named} | Brackets_Missing | Missing placeholders produce a warning which needs to be interpreted.
Missing%s and %(named)s | PercentS_Missing | Missing placeholders produce an error.
f-strings | N/A | They cannot be used as the placeholders would already be replaced by their final value before the translation kicks-in.

### Next Steps

The proposed solution was discussed and agreed in Core Team meetings, however no changes will be made to the codebase until we have a way to enforce the style.

Therefore, the next step is to work on some sort of linter (run as either an actual linter or a test in the test suite) to enforce a single string formatting mechanism across Python and HTML files.

Only then we can open a PR to fix the existing source strings.

### WARNING

Changing placeholder would make old translation strings invalid. Therefore, we should wait for the beginning of a release cycle (maybe a week or two after a new release) and notify translators immediately in order to give them as much time as possible to go through the now broken/missing translations.

### PR DESCRIPTION
**Pull request by loicteixeira**
Fix #7111. As discussed there, the chosen style is the old style formatting.

25a83e0572227259cca0dbaa441ef5b1f5a9de10 setups and configures [Semgrep](https://semgrep.dev/docs/getting-started/) in order to lint the source strings and prevent regressions.
It might not be perfect just yet, maybe there's false negative, but at least I didn't see any false positives (which we could always temporarily ignore with `# nosemgrep: id-of-the-error` until the rules are updated).

11e86ea14b18b4856a9975298e389efdfe0dd4a0 and 4c7106b6e84d60609a3f8754e5bb265f96619e0e (the second one deals with strings in JS, so it was separated to another commit for ease of review.) update the source strings to the desired format.

See a [compiled list of source string changes](https://gist.github.com/loicteixeira/bcc5e678e3d06af07bdb3afe9d149575#file-source_strings-diff).

5db259ecb2c00a5e20a3b17012622cf38a2d658e is unrelated, but since flake8 was running on the file, it would complain about it.

Now that's about 150 strings which translators would have to update as I don't think the auto matching whatever will be able to pick this up. In order to reduce the burden as much as possible, I've made the following script which will update the PO files with the new placeholders when possible.

See the [update script](https://gist.github.com/loicteixeira/bcc5e678e3d06af07bdb3afe9d149575#file-source_strings_udpate-py).

If all goes well, it would only leave the translators 4 translations to update (see the `AMBIGUOUS_REPLACEMENTS` constant in the script).

However I'm not 100% sure how to proceed, so here is what I'm thinking so far:

1. Merge this PR **OR** do it on this branch until everything is confirmed to work?
2. [Fetch translations](https://github.com/wagtail/wagtail/wiki/Managing-Wagtail-translations#fetching-new-translations-from-transifex)
3. Commit & push to `main` to create a "restore" point, just in case
4. Run the script `cd wagtail_root; python ./scripts/issue_7111_update_i18n_placeholders.py`
5. Commit
6. [Rebuild sources and push to Transifex](https://github.com/wagtail/wagtail/wiki/Managing-Wagtail-translations#generating-new-source-files-for-translation)
7. Push the translations with `tx push -t`

Unfortunately I can't really test it without actually doing it...

The main idea is that we can't push the updated source strings first because the translations correspondance would be lost, so we fetch first (step 2), update locally (step 4), push the updated source string (step 6), at this point the translations correspondance would be lost, but pushing the updated translations (step 7) should restore it 🤞 

The other question that I have is when pulling translations with the `fetch-translations.sh` script, we would normally [do some cleanup](https://github.com/wagtail/wagtail/blob/398c57515311b37f353c96e78231992b78bf3f19/scripts/fetch-translations.sh#L12) and maybe we don't want to do that (since we're going to push those files later), so we should only run the first 2 statements of the script (i.e. remove old translation files then pull).

---

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: ~Have you added tests to~ Existing tests cover the new/fixed behaviour?
-   [ ] ~For front-end changes: Did you test on all of Wagtail’s supported environments?~[^2]
    -   [ ] ~**Please list the exact browser and operating system versions you tested**:~
    -   [ ] ~**Please list which assistive technologies [^3] you tested**:~
-   [ ] ~For new features: Has the documentation been updated accordingly?~

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: .semgrep.yml
Comment (lb-) [inline .semgrep.yml:15]: I think we should really point the developer to a docs page here, hopefully I can get https://github.com/wagtail/wagtail/pull/9431 in and we can have this end docs file as a source of truth here.

The nuances of the various types of string formatting was a bit for me to understand, and there is the temptation to use the 'newer' thing. Having the docs as a place where can explain this a bit better would be useful.

Also, this means adding a docs section also I think as part of this PR.

**Review by lb- — COMMENTED** [review]
@loicteixeira this is incredible - the code looks solid to me, have not run locally yet though.

A few things
* We probably need eyes on this from a few more core team to get feedback (I will add the needs decision label and flag this for tomorrow's meeting)
* Adding `semgrep` as another dependency is probably required, but just wondering if this can be somehow written into a tool we already use like a flake8 plugin (note: I have done zero research into this, so feel free to shut that idea down). https://flake8.pycqa.org/en/latest/plugin-development/index.html
* Semgrep's licence looks like it's compatible with Wagtail - https://github.com/returntocorp/semgrep/blob/develop/LICENSE

File: .semgrep.yml
Comment (loicteixeira) [reply .semgrep.yml:15]: Yes, now that #9431 is merged, I'll rebase my branch tomorrow and piggyback on it 👍

**Comment by loicteixeira** [comment]
> Adding semgrep as another dependency is probably required, but just wondering if this can be somehow written into a tool we already use like a flake8 plugin (note: I have done zero research into this, so feel free to shut that idea down). [flake8.pycqa.org/en/latest/plugin-development/index.html](https://flake8.pycqa.org/en/latest/plugin-development/index.html)

Yes, it should be possible to do the same thing with a flake8 plugin, but it's a bit more involved since it works with the AST tree (where semgrep is some pseudo-regex) and would need to be published to pypi. I actually started exploring this, but finally switched to semgrep so I could get this out quickly and finally provide some much needed help to the translators.

I'd understand though if we don't want to introduce yet another dev dependency, but I'd be in favour of going with it for now, and I can try in a second time to replace semgrep with a flake8 plugin.

As a side note, merging the string changes without the linter is also possible, but obviously we would have to do this all over again later 😒

**Comment by loicteixeira** [comment]
Also, we probably need to wait for 4.1 to be released, otherwise we will mess up the translations for that release.

**Comment by lb-** [comment]
I think that reasoning makes sense - regarding the added dependency. I also think this should be merged as one, the fixes and the enforcing of the new approach.

Hopefully though there will be some other thoughts on this from others in the core team soon.

Thanks Loic.

**Comment by lb-** [comment]
Discussed at the core team last week, no major objections but a few refinements to do before this goes in. But now that 4.1 is out this can be revisited.

* @thibaudcolas to review
* Revisit how we can minimise the amount of changes needed in Transifex
* Did not seem like any major objections to the new dev dependency
* Might be good to get some feedback if possible from translators

Hopefully that summarises it ok.

Thanks for the epic progress on this @loicteixeira

**Comment by loicteixeira** [comment]
Rebased on `main` after the release of v4.1. The list of diff and the script in the PR description have been updated accordingly.

**Comment by loicteixeira** [comment]
Btw, if approved, I'm obviously happy to merge this myself and run the scripts since this isn't your usual merge & push.

**Comment by th3hamm0r** [comment]
@loicteixeira epic work, this really improves the quality of wagtail's i18n!

I've focused on the changed translation calls in python and the resulting strings (original and "de"), I've also tested your script locally.
I've found one issue with the script, there is at least one replacement, which is incomplete, therefore it does not work:

```python
# => Are you sure you want to delete this %(object)s? If other things in your
    "Are you sure you want to delete this %s? If other things in your": (
        ("%s", "%(object)s"),
    ),
```
Somehow you've only used parts of the original string. This finding was lucky, because the string hasn't been changed, but the (newer?) polib-version reformatted the string, so I've found the remaining "%s".
I haven't checked all other replacements yet, as there are some of them :grin: 

As you've pointed out above, not everything can be tested, but from my point of view it looks good :pray:

**Review by th3hamm0r — CHANGES_REQUESTED** [review]
Sry for spamming you with comments, my process was a bit iterative, I should have waited and used the review comment once :grimacing:

**Comment by loicteixeira** [comment]
No worries. Thank you for the thorough review 👍

**Comment by loicteixeira** [comment]
Rebased on main.

> Truncated "Are you sure you want to delete this %s?"

Nice catch. I've updated the script accordingly!

> I don't know exactly why it has been missed with semgrep

The semgrep was looking for a function called `_` or `gettext` or `gettext_lazy` or `ngettext`, but with only a single argument.

- d9be05d updates the pattern, so it also looks for more arguments (it still seems to work when there's only one)
- 2224fc8 adds `ngettext_lazy` to the list of functions to watch

In the case of ngettext, it's still not perfect as it will only be looking at the first of the two strings, but maybe that's good enough for a first iteration 🤔

> Also there are some uses of {...} in the admin's JS

The rule only looks for Python files. I'll have a look at enabling it for js/jsx/ts/tsx files as well, but I'll need to make sure it doesn't break anything.

**Comment by th3hamm0r** [comment]
Awesome!

> The rule only looks for Python files. I'll have a look at enabling it for js/jsx/ts/tsx files as well, but I'll need to make sure it doesn't break anything (regarding the replace/interpolate you mentioned).

Yes, I assumed that. Since I haven't used them yet, I cannot really give advice here, so for me it is more a nice to have.

File: wagtail/images/fields.py
Comment (gasman) [inline wagtail/images/fields.py:54]: Slightly expanding the scope of the task here, but it would be great if we could eliminate the hacky two-pass string formatting that's going on here (and the file_too_many_pixels message immediately below - I think these two are the only occurrences of this in the codebase).

Since Transifex can't recognise the `%%`-escaped variable as a placeholder (or at least, it couldn't around the time of 4.0 - I didn't see the problem recur in 4.1), we've had multiple cases of user error where translators have done things like inserting whitespace between the `%%` and `s`. (And, of course, adding the variable name here will be one more thing they can potentially get wrong...)

I think this can be fixed by leaving out the `% {"max_filesize": max_upload_size_text}` step here, and inserting that at the same time as `file_size`.

**Review by gasman — COMMENTED** [review]
Great work on this @loicteixeira!

**Comment by th3hamm0r** [comment]
Should we add `semgrep --config .semgrep.yml --error` also to the Makefile's `lint-server` section?

The PR's template guides everyone to execute `make lint`, so this would then also cover semgrep. I also use it quite often :grin:

**Comment by gasman** [comment]
> Should we add `semgrep --config .semgrep.yml --error` also to the Makefile's `lint-server` section?
> 
> The PR's template guides everyone to execute `make lint`, so this would then also cover semgrep. I also use it quite often 😁

Good idea! And the [CircleCI checks](https://github.com/wagtail/wagtail/blob/main/.circleci/config.yml) too.

**Comment by loicteixeira** [comment]
Ah of course! I'm used to projects running pre-commit in CI, so updating the pre-commit config is enough.

**Comment by th3hamm0r** [comment]
Big :+1: @loicteixeira 
After your changes, I've now tried a different approach: I've executed your replacement script, then I've used a modified version of "rebuild-translation-sources.sh" (using "de" instead of "en", and no initial deletion of the source files), so it was easier to review the output, which allowed me to find some other issues, which I totally understand, there are soo many changed strings :sweat_smile: :

1) Missing replacements:
```
    # => The page '%(page_title)s' is not currently awaiting moderation in task '%(task_name)s'.
    "The page '{0}' is not currently awaiting moderation in task '{1}'.": (
        ("{0}", "%(page_title)s"),
    ),
```
It is missing a second tuple `("{1}", "%(task_name)s"),`

```
    # => The page '%(page_title)s' was successfully created in %(locales)s
    "The page '{page_title}' was successfully created in {locales}": (
        ("{page_title}", "%(page_title)s"),
    ),
```
It is missing a second tuple `("{locales}", "%(locales)s"),`

2) Atm we have 3 replacements of the string `"Edit this %s"`, all with different placeholders. This leads to invalid strings, eg. for "de" the result is:
```
msgid "Edit this %(object_name)s"
msgstr "%(objects)s bearbeiten"
```

I think one replacement is invalid and can be removed(?):
```
    # => Edit this %(object)s
    "Edit this %s": (("%s", "%(objects)s"),),
```
It could be that it was previously used and you've unified it by changing it to "object", but forgot the lines in the replacement script.

For the remaining two:
What if we replace the placeholder `%(object_name)` with `%(object)`? This seems to be used by the majority of the python calls where the model's `verbose_name` is passed (documents, images, snippets,...).
Then we could remove this replacement too:
```
    # => Edit this %(object_name)s
    "Edit this %s": (("%s", "%(object_name)s"),),
```

3) The replacement for the this plurarl message does not work properly:
```
msgid "Include subtree (%(descendant_count)s page)"
msgid_plural "Include subtree (%(descendant_count)s pages)"
msgstr[0] "Teilbaum ({} Seite) einbeziehen"
msgstr[1] "Teilbaum ({} Seiten) einbeziehen"
```
This has been added after my comment https://github.com/wagtail/wagtail/pull/9377#issuecomment-1324790044, so its partially my fault :D
I think the line `entry.msgstr = entry.msgstr.replace(old, new, 1)` does not work here.

4) I think we cannot automate the replacement of the following, but I'm not completely sure:
```
    # => Not a valid .%(extension)s image. The extension does not match the file format (%(image_format)s)
    "Not a valid %s image.": (
        ("%s", "%(extension)s"),
        # The string was modified to have a second placeholder, so this little hack
        # hooks on the end of the string to insert the second part with the second placeholder.
        (
            " image.",
            " image. The extension does not match the file format (%(image_format)s)",
        ),
    ),
```
It only replaces the message id, because the language's string will probably never be replaced:
```
msgid ""
"Not a valid .%(extension)s image. The extension does not match the file "
"format (%(image_format)s)"
msgstr "Kein gültiges %(extension)s Bild."
```
I don't know if transifex will mark that string as untranslated then? If not, it would be easy to miss by translators.

5) `%(object)s` should be `%(image_title)s`:
```
    # => Image '%(image_title)s' updated.
    "Image '{0}' updated.": (("{0}", "%(object)s"),),
    # => Image '%(image_title)s' deleted.
    "Image '{0}' deleted.": (("{0}", "%(object)s"),),
    # => Image '%(image_title)s' added.
    "Image '{0}' added.": (("{0}", "%(object)s"),),
```

6) "objects" vs. "object"
```
    # => Collection '%(object)s' created.
    "Collection '{0}' created.": (("{0}", "%(objects)s"),),
    # => Collection '%(object)s' updated.
    "Collection '{0}' updated.": (("{0}", "%(objects)s"),),
    # => Collection '%(object)s' deleted.
    "Collection '{0}' deleted.": (("{0}", "%(objects)s"),),
...
    # => Choose %(object)s
    "Choose %s": (("%s", "%(objects)s"),),
    # => Choose another %(object)s
    "Choose another %s": (("%s", "%(objects)s"),),
    # => Edit this %(object)s
    "Edit this %s": (("%s", "%(objects)s"),),
    # => Group '%(object)s' created.
    "Group '{0}' created.": (("{0}", "%(objects)s"),),
    # => Group '%(object)s' updated.
    "Group '{0}' updated.": (("{0}", "%(objects)s"),),
    # => Group '%(object)s' deleted.
    "Group '{0}' deleted.": (("{0}", "%(objects)s"),),
    # => User '%(object)s' updated.
    "User '{0}' updated.": (("{0}", "%(objects)s"),),
    # => Editing %(object)s
    "Editing %s": (("%s", "%(objects)s"),),
    # => User '%(object)s' deleted.
    "User '{0}' deleted.": (("{0}", "%(objects)s"),),
```
The tuples have to use "object".

**Comment by th3hamm0r** [comment]
Maybe a little bit late, because I think the whole thing gets quite solid with the next iteration, but we could use [github-gists](https://gist.github.com) for the replacement script, so reviewing changes is easier (e.g. with direct-links to lines)?

**Comment by loicteixeira** [comment]
Thanks again for the thorough review!

And yes, gist would be a great place for the scripts. Even if this is almost at the finish line, I'll upload it there.

**Comment by loicteixeira** [comment]
> 1) Missing replacements
> 3) The replacement for the this plural message does not work properly
> 5) `%(object)s` should be `%(image_title)s`
> 6) "objects" vs. "object"

Script updated, see the [changes](https://gist.github.com/loicteixeira/bcc5e678e3d06af07bdb3afe9d149575/revisions).

> 2) Atm we have 3 replacements of the string `"Edit this %s"`, all with different placeholders. This leads to invalid strings, eg. for "de" the result is:

All updated to use `object`. I also updated related strings (add/delete this ...) so it use the same placeholder everywhere.

> 4) I think we cannot automate the replacement of the following, but I'm not completely sure

Indeed, this won't work. Ignoring it then and it will be up to the translators to retranslate.

**Review by th3hamm0r — CHANGES_REQUESTED** [review]
Nice!
I've rerun the scripts again and there seems to be only one issue with the new plural handling in your script. It produced the following diff:
```
-msgid "Include subtree ({} page)"
+msgid "Include subtree (%(descendant_count)s page)"
 msgid_plural "Include subtree ({} pages)"
-msgstr[0] "Teilbaum ({} Seite) einbeziehen"
-msgstr[1] "Teilbaum ({} Seiten) einbeziehen"
+msgstr[0] "Teilbaum (%(descendant_count)s Seite) einbeziehen"
+msgstr[1] "Teilbaum (%(descendant_count)s Seiten) einbeziehen"
```
The "msgid_plural" isn't replaced, so I think this has to be done in this block: https://gist.github.com/loicteixeira/bcc5e678e3d06af07bdb3afe9d149575#file-source_strings_udpate-py-L452-L454

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:406]: Since we're replacing these strings anyway, could you please also edit the strings in the snippets views to use more generic variable names?

Specifically, I need all occurrences of `snippet_type` to be renamed to something else e.g. `model_name`:
 https://github.com/wagtail/wagtail/blob/ef8a9702614448edaf1ee66bbbd13b42fc004dab/wagtail/snippets/views/snippets.py#L245-L257

Then, it's up to you to use `object` or `instance`, but make sure to be consistent for both views.

Once we've renamed the snippet-specific variables, we can use those strings in the generic views (they're more informative as they also provide the model name).

I think `snippet_type` is the only variable in the snippets views that is snippet-specific, but if you found any other ones please also rename them to be more generic. Thanks!

**Review by laymonage — COMMENTED** [review]
I love this PR, thank you very much for the great work!

I'm working on some refactoring between snippets and generic views in #9709, and I'd like to rename some variables in translatable strings so that both views can share the same strings. I was hoping to piggyback some strings in the snippets views, if you don't mind? Thanks (and sorry for adding more work)!

File: wagtail/admin/views/generic/models.py
Comment (lb-) [reply wagtail/admin/views/generic/models.py:406]: Maybe we can do this rename as a follow up PR? Might be good to land this in `main` soon. This way new PRs can come in with the linting rules.

Comment (loicteixeira) [reply wagtail/admin/views/generic/models.py:406]: I'll merge in the next hour or two. If there aren't too many changes for snippets, I'll try to do it at the same time. Otherwise, I'll leave it to a follow-up PR indeed.

Comment (loicteixeira) [reply wagtail/admin/views/generic/models.py:406]: Done in 6d9476d with some additional normalization in 868c239.
Corresponding [diffs update](https://gist.github.com/loicteixeira/bcc5e678e3d06af07bdb3afe9d149575/revisions#diff-b6e4f31103721ae34912b0ec3248c72d9054c3db48c3e40b97d8129cb328d8fd) and [script update](https://gist.github.com/loicteixeira/bcc5e678e3d06af07bdb3afe9d149575/revisions#diff-c639433905560e28f8d413989b641d1b2b4773333faa845142da87b8a08275eb).

Onto the merge and Transifex update now 🤞

Comment (lb-) [reply wagtail/admin/views/generic/models.py:406]: Thanks Loic. You are a legend

**Review by th3hamm0r — APPROVED** [review]
I've reran the scripts, looks good! Again, amazing work @loicteixeira and all contributors!

I think we should keep your script (or better you as the "polib-pro") in mind, because changing placeholder names during refactorings will probably happen again in the future, and this little script avoids creating unnecessary (maybe release-blocking) tasks for the translators :+1:

**Comment by loicteixeira** [comment]
I merged locally, but the re-upload to Transifex didn't really go as planned. Trying to fix it and will update here with the exact steps taken.

**Comment by loicteixeira** [comment]
Steps taken to merge this for posterity, and also because I'll have to do it all over again for wagtail-localize which is in the same Transifex project.

```bash
BRANCH_NAME="7111-enforce-translation-string-formatting"

# Merge
git switch main
git pull --ff-only upstream main
git switch "${BRANCH_NAME}"
git reset $(git merge-base main "${BRANCH_NAME}")
git add .
git commit -m "Enforce the use of a single string formatting mechanism for translation source strings" -m "Close #9377"
git rebase main
git merge --ff-only "${BRANCH_NAME}"

# Pull & commit existing translations
cd scripts
./fetch-translations.sh > ../../logs/translation.out 2>../../logs/translation.err
python ./check-translation-strings.py
cd ..
git add .
git commit -m "Fetch new translations from Transifex"

# Update the strings and sanity check with French translation (I can't possibly check all languages...)
python ./scripts/source_strings_update.py
find wagtail -name *.po -type f -path **/fr/** -exec git diff --staged --word-diff {} \;
git add .

# Rebuild English translations to check if there's diff from what the script has done
cd scripts
./rebuild-translation-sources.sh
cd ..
git add .

git commit -m "Update translations placeholder names" -m "See 364de3d4dc and #9377"

# Last linting check
semgrep --config .semgrep.yml --error .
```

Now the tricky part:

```
# Upload English source strings to Transifex
tx push --source
```

Languages now show 160 untranslated strings

```
# Upload French translations to Transifex for sanity check on Transifex
tx push --translations --language=fr
```

Sadly it did not work, files were being "skipped", and we still had 160 strings to translate.

Trying to fix it:

```bash
git checkout HEAD~1  # Come back to the commit before the placeholder update
tx push --source  # Re-push the original English sources
```

At this point Transifex is back to what it was before I started.

```bash
git switch main

# Delete translations
find wagtail -iname *.po ! -iwholename */en/* -delete

# Re-fetch, but without using `scripts/fetch-translations.sh` to keep strings location information.
tx pull --all --minimum-perc=1

# Re-update placeholders
python ./scripts/source_strings_update.py

# Re-upload English updated source strings to Transifex
tx push --source

# Re-upload French translations
tx push --translations --language=fr
```

Unfortunately files are still being "skipped".
While the location information "might" be required anyway, the files seem to be skipped before of modification date.
Trying to force the upload.

```bash
# Re-upload French translations
tx push --translations --language=fr --force --no-interactive
```

This work, but it means that the time between fetch and upload has to be as short as possible
or we will overwrite work from translations which might be updating strings at the same time we do this work.

```bash
# Upload other translations
tx push --translations --force --no-interactive
```

So at this point I'm not sure if files were skipped because of the missing location information or simply the timestamps.
It feels like it would be the latter (location information shouldn't be needed as long as you have the string "id")
but I don't want to fiddle with it anymore for today.

In the end, there are:

- 10 strings which weren't updated (that's 5 more than the 5 impossible/ambiguous auto-update we had identified).
- [4 strings which cannot use the percent formatting](https://github.com/wagtail/wagtail/blob/56d0fe8fbc56737c0d7729f1a32ae61a2cac6f1e/wagtail/locks.py#L47-L79) were flagged as untranslated, I'm not sure why.

So 160 strings down to 14. Not perfect but not bad either.

```bash
# Cleanup & merge
git restore .
git push --dry-run upstream main
git push upstream main
```

So this might need some refinement before we start using it on a "regular" basis.

a cartoon dog sitting at a table in the middle of a room in fire saying "this is fine" and looking into its coffee cup

Anyway, thank you @lb-, @th3hamm0r, @gasman & @laymonage for your help on this issue.

**Comment by lb-** [comment]
Wow @loicteixeira - thanks for documenting this process also.

**Pull request closed without merging**
